shell/platform/linux/fl_keyboard_manager.cc

```diff
@@ -296,20 +257,46 @@ static void responder_handle_event_callback(bool handled,
                          event_type == GDK_KEY_RELEASE);
         gdk_event_put(fl_key_event_get_origin(event));
       }
-    } else {
-      g_object_unref(pending);
     }
   }
 }
 
 static void responder_handle_embedder_event_callback(bool handled,
-                                                     gpointer user_data_ptr) {
-  responder_handle_event_callback(handled, user_data_ptr, TRUE);
+                                                     gpointer user_data) {
+  g_autoptr(FlKeyboardManagerData) data = FL_KEYBOARD_MANAGER_DATA(user_data);
+
+  fl_keyboard_pending_event_mark_embedder_replied(data->pending, handled);
+
+  g_autoptr(FlKeyboardManager) self =
+      FL_KEYBOARD_MANAGER(g_weak_ref_get(&data->manager));
+  if (self == nullptr) {
+    return;
+  }
+
+  responder_handle_event_callback(self, data->pending);
 }
 
-static void responder_handle_channel_event_callback(bool handled,
-                                                    gpointer user_data_ptr) {
-  responder_handle_event_callback(handled, user_data_ptr, FALSE);
+static void responder_handle_channel_event_cb(GObject* object,
+                                              GAsyncResult* result,
+                                              gpointer user_data) {
+  g_autoptr(FlKeyboardManagerData) data = FL_KEYBOARD_MANAGER_DATA(user_data);
+
+  g_autoptr(FlKeyboardManager) self =
+      FL_KEYBOARD_MANAGER(g_weak_ref_get(&data->manager));
+  if (self == nullptr) {
+    return;
```

Comment:
No, the engine is a weak pointer, so it may be destroyed before the manager is. In this case, there's nothing to do, so it makes sense to complete at this point. In normal behaviour, both objects will be destroyed at the same time; the weak pointer is there to stop a reference loop. There might be a call in progress, so if that is handled after the engine is destroyed this stops it accessing freed memory.

This has prompted me to move the getting of the weak reference until after the finish is called, as then a cancelled async call won't attempt to do anything. I'm pretty sure this weak reference will be removed in a future refactoring however.